modules/sdk/src/test/scala/org/tessellation/sdk/domain/cluster/programs/PeerDiscoverySuite.scala

```diff
@@ -0,0 +1,126 @@
+        IO(discoverdPeers.map(p => p))
+      }
+
+    }
+
+    val clusterStorage = new ClusterStorage[IO] {
+
+      override def getPeers: IO[Set[Peer]] = IO(peersInCluster)
+
+      override def getPeer(id: PeerId): IO[Option[Peer]] = ???
+
+      override def addPeer(peer: Peer): IO[Unit] = ???
+
+      override def hasPeerId(id: PeerId): IO[Boolean] = ???
+
+      override def hasPeerHostPort(host: Host, p2pPort: Port): IO[Boolean] = ???
+
+      override def setPeerState(id: PeerId, state: node.NodeState): IO[Unit] = ???
+
+      override def removePeer(id: PeerId): IO[Unit] = ???
+
+      override def removePeers(ids: Set[PeerId]): IO[Unit] = ???
+
+      override def peerChanges: fs2.Stream[IO, (PeerId, Option[Peer])] = ???
+
+      override def setClusterSession(token: cluster.ClusterSessionToken): IO[Unit] = ???
+
+      override def getClusterSession: IO[Option[cluster.ClusterSessionToken]] = ???
+
+      override def getClusterId: cluster.ClusterId = ???
+
+    }
+
+    PeerDiscovery.make(clusterClient, clusterStorage, nodeId)
+  }
+
+  test("discoverFrom - only unknown peers should be returned") {
+    val gen = for {
+      commonPeers <- Gen.containerOf[Set, Peer](peerGen)
+      newPeers <- Gen.containerOf[Set, Peer](peerGen)
+      discoverFromPeer <- peerGen
+      selfPeer <- peerGen.map(_.copy(id = nodeId))
+      knownPeers <- Gen.containerOf[Set, Peer](peerGen)
+    } yield (discoverFromPeer, selfPeer, commonPeers, newPeers, knownPeers)
+
+    forall(gen) {
+      case (discoverFromPeer, selfPeer, commonPeers, newPeers, knownPeers) =>
+        for {
+          peerDiscovery <- mkPeerDiscovery(commonPeers ++ knownPeers, Set(discoverFromPeer, selfPeer) ++ commonPeers ++ newPeers)
+          res <- peerDiscovery.discoverFrom(discoverFromPeer)
+        } yield expect.same(res, newPeers.map[P2PContext](identity))
+    }
+  }
+
+  test("discoverFrom - no new peers should be returned when all has been seen already") {
+    val gen = for {
+      commonPeers <- Gen.containerOf[Set, Peer](peerGen)
+      newPeers <- Gen.containerOf[Set, Peer](peerGen)
+      knownPeers <- Gen.containerOf[Set, Peer](peerGen)
+    } yield (commonPeers, newPeers, knownPeers)
+
+    forall(gen) {
+      case (commonPeers, newPeers, knownPeers) =>
+        for {
+          peerDiscovery <- mkPeerDiscovery(commonPeers ++ knownPeers, commonPeers ++ newPeers)
+          _ <- peerDiscovery.discoverFrom(peerGen.sample.get)
+          res <- peerDiscovery.discoverFrom(peerGen.sample.get)
+        } yield expect.same(res, Set.empty)
+    }
+  }
+
+  test("getPeers - only peers not saved in cluster should be cached") {
+    val gen = for {
+      commonPeers <- Gen.containerOf[Set, Peer](peerGen)
+      newPeers <- Gen.containerOf[Set, Peer](peerGen)
+      discoverFromPeer <- peerGen
+      selfPeer <- peerGen.map(_.copy(id = nodeId))
+      knownPeers <- Gen.containerOf[Set, Peer](peerGen)
```

Comment:
I see this gen is duplicated so it can be extracted (IMO second test case can omit selfPeer and discoverFromPeer, so you can reuse it for all cases)